### ISSUE TITLE
 Build problems with gcc 10 on linux

### ISSUE THREAD

**Issue by biagas** [visit-dav/live-customer-response#216]
We're trying to [combine `visit-users` email list with GitHub issues](https://github.com/visit-dav/live-customer-response/wiki/How-the-new-GitHub-visit-users-Email-Integration-Works).
When replying on `visit-users`, please reply **only** to emails with a *Subject*
line that includes `[visit-dav/live-customer-response]`.

---
```
Hello,
I've been trying to build VisIt 3.0 and 3.1 on a linux cluster and hit a number of snags.
Some I was able to fix:
•  The renaming of mesa to mesagl is missing in some environment variables in build_visit3_1_4 at least.
•  The python build configuration tries to add Intel flags if the compiler name contains *icc*
    which fails with the mpicc wrapper (with gcc).
•  I had to apply the Qt gcc 9.2 patch also with gcc 10.2 and I had to add that manually as 
   the check didn't succeed with the mpi wrappers. When I tried building Visit 3.0.2 with gcc 10 I
   copied the patch function from the build_visit_3_1_4 script.
•  I had to add some flags manually for the python build setup to find libffi, tcl and tk on the 
   system but those might have been difficult to find due to the module system, I'm not sure.
•  The intrinsics declared at nano mesa-17.2.8/src/gallium/drivers/swr/rasterizer/common/os.h:171 
    and following exist in gcc 9.3 and 10.2 so I commented out these lines.

With the above fixes I was able to build a working VisIt 3.1.4 with gcc 9.3.0, however I suspect the MPI
version I have available on that cluster built with that gcc has issues. So I am trying to build VisIt 3.0.2 
or 3.1.4 with a newer MPI version which is built with gcc 10.2.0 and there I hit the below issue when linking 
bin/vtkpython during the VTK build.
===
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkUnicodeString::from_utf8(char const*)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkUnicodeString::operator=(vtkUnicodeString const&)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkUnicodeString::utf8_str(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) const'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkObjectBase::vtkObjectBase()'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkObjectBase::InitializeObjectBase()'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkCommand::UnRegister()'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkObjectBase::IsTypeOf(char const*)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkObjectBase::~vtkObjectBase()'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkCommand::GetStringFromEventId(unsigned long)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkCommand::vtkCommand()'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'typeinfo for vtkObjectBase'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkObjectBase::Delete()'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkObjectBase::FastDelete()'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkObjectBase::PrintSelf(std::ostream&, vtkIndent)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkObjectBase::PrintHeader(std::ostream&, vtkIndent)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkObjectBase::PrintTrailer(std::ostream&, vtkIndent)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkObjectBase::Register(vtkObjectBase*)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkObjectBase::UnRegisterInternal(vtkObjectBase*, int)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkObjectBase::ReportReferences(vtkGarbageCollector*)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkObject::GetGlobalWarningDisplay()'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkOStrStreamWrapper::vtkOStrStreamWrapper()'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkOStreamWrapper::operator<<(char const*)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkOStreamWrapper::operator<<(int)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkOStreamWrapper::operator<<(char)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkOStrStreamWrapper::str()'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkOutputWindowDisplayGenericWarningText(char const*)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkOStrStreamWrapper::rdbuf()'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkOStrStreamWrapper::freeze(int)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkOStrStreamWrapper::~vtkOStrStreamWrapper()'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkWeakPointerBase::~vtkWeakPointerBase()'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkWeakPointerBase::~vtkWeakPointerBase()'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkWeakPointerBase::operator=(vtkWeakPointerBase const&)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkVariant::GetType() const'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkVariant::ToString() const'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkVariant::ToVTKObject() const'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkVariant::ToUnicodeString() const'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkUnicodeString::utf8_str() const'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkObjectBase::GetClassName() const'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkWeakPointerBase::vtkWeakPointerBase(vtkWeakPointerBase const&)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkWeakPointerBase::vtkWeakPointerBase(vtkObjectBase*)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkWeakPointerBase::operator=(vtkObjectBase*)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkObject::GetCommand(unsigned long)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkObjectBase::Print(std::ostream&)'
../../lib/libvtkWrappingPython27Core-8.1.so.1: error: undefined reference to 'vtkObjectBase::RegisterInternal(vtkObjectBase*, int)'
collect2: error: ld returned 1 exit status
make[2]: *** [bin/vtkpython] Error 1
make[1]: *** [Wrapping/Python/CMakeFiles/vtkpython.dir/all] Error 2
===
With a lot of digging I was able to link by hand by removing one of the libraries in the linking command, however then the installation fails due to a mismatch when it tries to change the rpath, and working around that eventually leads to an error in the building of visit proper.
So my question is: how do I fix this linking of vtkpython so that I can proceed?
Thanks in advance for your help,
Best regards,
Yann
```

**Comment by biagas** [comment]
Hello Yann,

Thank you for the report on the build_visit issues.
I have entered a ticket for the MESA env var problem.

As for the python issue, was this a problem within build_visit or python itself? If within python itself, we are moving to python3 for our next release, so this may already be resolved.

Internal mesa issue: we are also moving to a slightly newer version of mesa for our next release.

vtkPython linking:  not sure about that, will do some checking and get back with you.

**Comment by biagas** [comment]
Regarding vtk linking issues. Is there anything the build_visit log file that indicates a different error with vtkPythonWrapping? Perhaps the vtk build didn't get its python settings correct? If you can attach the build_visit log file it might be helpful in diagnosing the problem.
Thank you.

**Comment by ykempf** [comment]
Let's do it here then indeed.
[build_visit3_0_2_log.tar.gz](https://github.com/visit-dav/live-customer-response/files/5889712/build_visit3_0_2_log.tar.gz)

**Comment by biagas** [comment]
It looks like build_visit was using the mpicc compiler instead of gcc
```
C++ compiler on this system is: /appl/opt/OpenMPI/4.0.5-GCC-10.2.0/bin/mpicxx  -std=gnu++14 
```
The don't think mpicxx should be the compiler used when building third party libs or visit proper.
Is there a default gcc compiler you can use? Is it a module that needs to be loaded so that it is in your path?
You can try setting path to gcc/g++ via environment vars when calling build_visit, eg (in tcsh):
env CC=/path/to/gcc CXX=/path/to/g++ build_visit --all your build_visit args

**Comment by biagas** [comment]
You might find these detailed notes on [build_visit usage](https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/gui_manual/Building/index.html) in our manual to be helpful.

**Comment by ykempf** [comment]
I've always used the MPI wrappers to build on a variety of architectures (for almost a decade now) but I'll admit that I have been reusing "what worked before" for almost as long. I wasn't aware of this manual, I'll look through and see what I should do better. A quick look now seems to indicate that I should use the wrappers only for the parallel parts and not everything indeed, so I'll try that. Although mpicxx is really just the wrapper for g++.

**Comment by ykempf** [comment]
I now try to use the flags below. I switched the ones which are not explicitly for MPI or parallel settings to gcc/g++ instead of mpicc and mpicxx. However it now fails to build numpy already, it's lacking math headers and libraries, which are probably set by the wrapper but not when using gcc/g++ straight.

I'll check next week which of these I should revert to proceed.

```
export MPI_CXX_COMPILER=/appl/opt/OpenMPI/4.0.5-GCC-10.2.0/bin/mpicxx
export MPI_C_COMPILER=/appl/opt/OpenMPI/4.0.5-GCC-10.2.0/bin/mpicc
export MPI_C=/appl/opt/OpenMPI/4.0.5-GCC-10.2.0/bin/mpicc

export CMAKE_CXX_COMPILER=/appl/opt/GCCcore/10.2.0/bin/g++
export CMAKE_C_COMPILER=/appl/opt/GCCcore/10.2.0/bin/gcc

export PAR_COMPILER=/appl/opt/OpenMPI/4.0.5-GCC-10.2.0/bin/mpicc
export PAR_INCLUDE=-I/appl/opt/OpenMPI/4.0.5-GCC-10.2.0/include
export PAR_LIBS=/appl/opt/OpenMPI/4.0.5-GCC-10.2.0/lib

export CXX=/appl/opt/GCCcore/10.2.0/bin/g++
export CC=/appl/opt/GCCcore/10.2.0/bin/g++
export cc=/appl/opt/GCCcore/10.2.0/bin/gcc
export CXX_COMPILER=/appl/opt/GCCcore/10.2.0/bin/g++
export C_COMPILER=/appl/opt/GCCcore/10.2.0/bin/gcc
```

**Comment by biagas** [comment]
I'm really not sure how to advise at this point. We've had problems with mpicc in the past, which is why we recommend not using it to build the third party libraries.  I know that g++ sometimes adds link flags automatically that mpicc does not and vice versa.
It might be time to go back to your original problem linking VTK python wrappers and try to sort out which link flags are are being used, which are missing.

If your previous build attempt using mpi is still around, you can cd to the VTK build dir, and delve into the link.txt, flags.make and depend.make files located here:
VTK-8.1.0-build/Wrapping/PythonCore/CMakeFiles/vtkWrappingPythonCore.dir

**Comment by brugger1** [comment]
We had a report of numpy not building on OSX 10.15. Here is the proposed fix.

The issue is in: /Applications/VisIt.app/Contents/Resources/3.1.4/darwinx86_64/lib/python/lib/python2.7/ctypes/__init__.py
Need to comment out line 276 to fix the problem:

262 def _reset_cache():
263     _pointer_type_cache.clear()
264     _c_functype_cache.clear()
265     if _os.name in ("nt", "ce"):
266         _win_functype_cache.clear()
267     # _SimpleCData.c_wchar_p_from_param
268     POINTER(c_wchar).from_param = c_wchar_p.from_param
269     # _SimpleCData.c_char_p_from_param
270     POINTER(c_char).from_param = c_char_p.from_param
271     _pointer_type_cache[None] = c_void_p
272     # XXX for whatever reasons, creating the first instance of a callback
273     # function is needed for the unittests on Win64 to succeed.  This MAY
274     # be a compiler bug, since the problem occurs only when _ctypes is
275     # compiled with the MS SDK compiler.  Or an uninitialized variable?
276     CFUNCTYPE(c_int)(lambda: None)

**Comment by ykempf** [comment]
Right, I got rid of the numpy issue, I had some problem with loading all the relevant modules and not cleaning properly before building again.

I managed to build and install everything adding the `--mpich` option and removing all the flags pointing to compilers. Somehow it managed to track down an mpicc nevertheless and I suspect it's a tangle of a system OpenMPI and built MPICH. In any case I didn't get the parallel engine to launch, it ran into MPI errors, but it is possible that SLURM is not cooperating with the MPICH I built.

Then I went back to the original linking issue. Now I am setting only the MPI_C* and PAR_* variable to point to the MPI wrappers and the others point to gcc/g++. I get the same vtkpython linking problem.

`link.txt` has:
```
/appl/opt/GCCcore/10.2.0/bin/g++ -fPIC -O2  -O3 -DNDEBUG  -Wl,-lc   -shared -Wl,-soname,libvtkWrappingPython27Core-8.1.so.1 -o ../../lib/libvtkWrappingPython27Core-8.1.so.1 CMakeFiles/vtkWrappingPythonCore.dir/vtkPythonArgs.cxx.o CMakeFiles/vtkWrappingPythonCore.dir/vtkPythonCommand.cxx.o CMakeFiles/vtkWrappingPythonCore.dir/vtkPythonOverload.cxx.o CMakeFiles/vtkWrappingPythonCore.dir/vtkPythonUtil.cxx.o CMakeFiles/vtkWrappingPythonCore.dir/vtkSmartPyObject.cxx.o CMakeFiles/vtkWrappingPythonCore.dir/PyVTKExtras.cxx.o CMakeFiles/vtkWrappingPythonCore.dir/PyVTKMethodDescriptor.cxx.o CMakeFiles/vtkWrappingPythonCore.dir/PyVTKReference.cxx.o CMakeFiles/vtkWrappingPythonCore.dir/PyVTKNamespace.cxx.o CMakeFiles/vtkWrappingPythonCore.dir/PyVTKObject.cxx.o CMakeFiles/vtkWrappingPythonCore.dir/PyVTKSpecialObject.cxx.o CMakeFiles/vtkWrappingPythonCore.dir/PyVTKTemplate.cxx.o -Wl,-rpath,/proj/ykempf/visit3.1.4-build-gcc10.2.0/VTK-8.1.0-build/lib: ../../lib/libvtkCommonCore-8.1.so.1 ../../lib/libvtksys-8.1.so.1 -ldl
```
and this does not complain if I run it on the command line.

`flags.make` contains
```
# CMAKE generated file: DO NOT EDIT!
# Generated by "Unix Makefiles" Generator, CMake Version 3.9

# compile CXX with /appl/opt/GCCcore/10.2.0/bin/g++
CXX_FLAGS = -O2  -O3 -DNDEBUG -fPIC -fvisibility=hidden   -std=c++11

CXX_DEFINES = -DVTK_IN_VTK -DvtkWrappingPythonCore_EXPORTS

CXX_INCLUDES = -I/proj/ykempf/visit3.1.4-build-gcc10.2.0/VTK-8.1.0/Utilities -I/proj/ykempf/visit3.1.4-build-gcc10.2.0/VTK-8.1.0-build/Wrapping/PythonCore -I/proj/ykempf/visit3.1.4-build-gcc10.2.0/VTK-8.1.0/Wrapping/PythonCore -I/proj/ykempf/visit3.1.4-build-gcc10.2.0/VTK-8.1.0-build/Common/Core -I/proj/ykempf/visit3.1.4-build-gcc10.2.0/VTK-8.1.0/Common/Core -I/proj/ykempf/visit3.1.4-build-gcc10.2.0/VTK-8.1.0-build/Utilities/KWIML -I/proj/ykempf/visit3.1.4-build-gcc10.2.0/VTK-8.1.0/Utilities/KWIML -I/proj/ykempf/visit3.1.4-build-gcc10.2.0/VTK-8.1.0-build/Utilities/KWSys -I/proj/ykempf/visit3.1.4-build-gcc10.2.0/VTK-8.1.0/Utilities/KWSys -I/proj/ykempf/visit3.1.4-build-gcc10.2.0/VTK-8.1.0-build/Utilities/Python -I/proj/ykempf/visit3.1.4-build-gcc10.2.0/VTK-8.1.0/Utilities/Python -I/proj/ykempf/visit3.1.4-build-gcc10.2.0/3rdparty/python/2.7.14/x86_64_gcc/include/python2.7 -I/proj/ykempf/visit3.1.4-build-gcc10.2.0/VTK-8.1.0-build/Wrapping/Tools -I/proj/ykempf/visit3.1.4-build-gcc10.2.0/VTK-8.1.0/Wrapping/Tools -I/proj/ykempf/visit3.1.4-build-gcc10.2.0/VTK-8.1.0-build/Utilities/KWSys/vtksys
```

and `depend.make` is over 1k lines long. I'll attach it with a txt extension.
[depend.make.txt](https://github.com/visit-dav/live-customer-response/files/5921754/depend.make.txt)

Let me know if you need more information. Thanks in advance for your help!

**Comment by ykempf** [comment]
I made some progress. I managed to get a binary package to run, so I am not continuing to try to fix the compilation with GCC 10.2.0. Also, it is not excluded that I had blunders in that build attempt...

With the binary package we stumbled upon the interesting issue that the engine_par runs fine with anything up to 8 processes, and beyond that it hangs. We also found out that the limit is 8 processes per node, i.e. we can run e.g. 128 processes on 16 nodes. The nodes we tested on have two different architectures, and both architectures have more then 8 cores, so we are stumped by this. We tested that it is independent of the data (we used silo files from the data directory as well as our own data written with thousands of processes), independent of the node architecture, independent of the file system (NFS and lustre) and we don't think it is an issue with SLURM. Yet any engine on more than 8 processes per node hangs until timeout. Is this number 8 something you have seen before? Would you have any hint as to what goes wrong?

**Comment by alhom** [comment]
I can add some more magic to this discussion, since I'm running Yann's build with 9 processors per node (currently doing a render with 16 nodes - 144 procs). 10 processors per node is the first to hang for me.

**Comment by ykempf** [comment]
I'll add that I just managed to complete a build of 3.1.4 from scratch using the `--mpich` option. And that has the same behaviour, hanging above 8 or 9 processes (per node).

**Comment by brugger1** [comment]
There's nothing special about 8 or 9 processes per node. On our clusters we run with 8, 16, 32 and 36 processors per node. Those are the number of cores per node on our various clusters. Things should work fine with any arbitrary processor count per node. That said, we have always used the system mpi on our clusters and not our mpich. How are you launching the parallel job (providing the output that VisIt outputs to the terminal would be useful to see)? Are you using the mpirun (or mpiexec) that is installed with VisIt or are you using a system one?

**Comment by ykempf** [comment]
It reports and launches
```
Running: sbatch --export=HOME=/home/ykempf,LIBPATH=/proj/ykempf/visit3.1.4-build-gcc9.3.0-mpich/3.1.4/linux-x86_64/lib,LD_LIBRARY_PATH=/proj/ykempf/visit3.1.4-build-gcc9.3.0-mpich/3.1.4/linux-x86_64/lib/osmesa:/proj/ykempf/visit3.1.4-build-gcc9.3.0-mpich/3.1.4/linux-x86_64/lib/mesagl:/proj/ykempf/visit3.1.4-build-gcc9.3.0-mpich/3.1.4/linux-x86_64/lib:/proj/ykempf/visit3.1.4-build-gcc9.3.0-mpich/bin/../current/linux-x86_64/lib:/appl/opt/cURL/7.69.1-GCCcore-9.3.0/lib:/appl/opt/bzip2/1.0.8-GCCcore-9.3.0/lib:/appl/opt/ncurses/6.2-GCCcore-9.3.0/lib:/appl/opt/binutils/2.34-GCCcore-9.3.0/lib:/appl/opt/zlib/1.2.11-GCCcore-9.3.0/lib:/appl/opt/GCCcore/9.3.0/lib/gcc/x86_64-pc-linux-gnu/9.3.0:/appl/opt/GCCcore/9.3.0/lib64:/appl/opt/GCCcore/9.3.0/lib,VISITHOME=/proj/ykempf/visit3.1.4-build-gcc9.3.0-mpich/3.1.4,VISITARCHHOME=/proj/ykempf/visit3.1.4-build-gcc9.3.0-mpich/3.1.4/linux-x86_64,VISITPLUGINDIR=/home/ykempf/.visit/3.1.4/linux-x86_64/plugins:/proj/ykempf/visit3.1.4-build-gcc9.3.0-mpich/3.1.4/linux-x86_64/plugins --mem=128G -M carrington --time=0:10:0 --ntasks=8 --nodes=1 --tasks-per-node=8 /tmp/visit.ykempf.15:55:24
```
and runs the job script containing this line
```
srun --export=ALL --mpi=pmi2 --ntasks=8 --nodes=1 --tasks-per-node=8 /proj/ykempf/visit3.1.4-build-gcc9.3.0-mpich/3.1.4/linux-x86_64/bin/engine_par -dir /proj/ykempf/visit3.1.4-build-gcc9.3.0-mpich -forcestatic -idle-timeout 480 -noloopback -sshtunneling -host turso02.cs.helsinki.fi -port 30665 -key dccd5db1a747645779d0
```

`srun` is the system's and it requires the mpi=pmi2 flag to handle mpich properly.

The issue is that I only managed to build against a system OpenMPI 4 and that seems to be buggy or unusably slow (not only with visit, also other software), that's why I went with the binary package and the `--mpich` flag.

One of our sysadmins was wondering if visit is trying to detect/adapt to the network/file system and therefore limiting the process count per node, as it might be beneficial in terms of per-node throughput.

**Comment by brugger1** [comment]
VisIt doesn't have any logic to limit the process count per node.

I did a little research and found some information at

https://slurm.schedmd.com/mpi_guide.html#mpich2

that mpich must be built with specific flags for it to work properly with slurm. It may be that you are getting somewhat lucky and things are working for the case of up to 8 or 9 tasks per node.

It specifically mentions setting --with-slurm, so this would most likely mean that it has to be compiled on the system you are building VisIt for. You could probably modify the build_visit script to add those flags to the mpich configure line.

That of course takes us back to the numpy issue. Can you send output from the numpy build failure? The other option is to wait a couple of weeks until we release 3.2, where we've moved to Python 3 and I suspect there won't be any issue with building numpy.

**Comment by ykempf** [comment]
Thanks for the hint, I am now trying two things, building everything with AOCC/clang as we built an OpenMPI with slurm optimisations and AMD optimisations for the AMD cluster; and building the version with MPICH using the --with-slurm option. I'll report on the results.

I don't have issues with numpy any more.

**Comment by brugger1** [comment]
@ykempf , did you make any more progress on this?

**Comment by ykempf** [comment]
Sorry for the silence. Indeed I eventually managed to build with AOCC/clang against an OpenMPI we had built with the same. I had to disable some optimisations that were crashing the compiler, among others, but I went through and runs with any number of processes I request. I now get a crash with our own plugin, which is most likely an issue with the plugin itself.

So this has drifted quite a bit from the original topic but I found a solution that will do for now.

Thanks for your help!

**Comment by brugger1** [comment]
Glad to hear you got things working!

**Comment by ykempf** [comment]
Hello, I fear I'll have to reopen this issue. The version of MPI I had used to build against triggered other bugs with the system. So I'm back with trying to build with GCC 10.2.0 and OpenMPI 4.0.5.

- I went to VisIt 3.2.0.
- I am using the system python 3.8.
- I went for server components only at the moment as I had issues with Qt/Qwt.
- I am using the MPI wrappers only for parallel things, and gcc/g++ for the rest.

And I get exactly the same vtkpython linking error above. Any hints how this could be fixed?

Also, I got compiler errors similar to https://github.com/tensorflow/tensorflow/issues/34197 so I had to hack into a handful of files, replacing `nullptr` on the `tp_print` lines with either `NULL` or `sizeof(Py_ssize_t)` until I got to the vtkpython linking error.
